Injections without pain

koder95 opened the pull request.

koder95 committed:
Add retention and target to annotations

Added @Retention(RUNTIME) and @Target to Component and Inject annotations to specify their applicable elements and runtime availability. This enables proper usage in dependency injection frameworks.

koder95 committed:
Add @Component and @Inject annotations to services

Annotated FileReaderServiceImpl, ProductParserImpl, and ProductServiceImpl with @Component for dependency injection. Added @Inject to ProductParser and FileReaderService fields in ProductServiceImpl to enable automatic wiring.

koder95 committed:
Implement basic dependency injection in Injector

Added logic to Injector for scanning packages, finding component classes, and instantiating dependencies using reflection. Updated Main to use Injector for ProductService instantiation. This enables automatic injection of dependencies annotated with @Inject.

mateacademy-ai-mentor commented on src/main/java/mate.academy/lib/Injector.java:
```diff
@@ -1,13 +1,128 @@
+        Field[] fields = implementation.getDeclaredFields();
+        Object o;
+        try {
+            o = implementation.getConstructor().newInstance();
```
This implementation creates instances unconditionally without consulting an instances cache. This violates the checklist item: "Let's check instances map before new instance creation." Consider adding a class-level Map<Class<?>, Object> instances cache, check it before creating a new instance, and store the created instance there to ensure one instance per type and help with circular dependencies.

mateacademy-ai-mentor commented on src/main/java/mate.academy/lib/Injector.java:
```diff
@@ -1,13 +1,128 @@
+    private static final Set<Class<?>> CLASS_SET = SELF_PACKAGES.stream()
+            .flatMap(s -> Injector.findAllClasses(s).stream())
+            .collect(Collectors.toSet());
+    private static final Set<Class<?>> COMPONENTS = CLASS_SET.stream()
```
The code resolves implementations by searching COMPONENTS rather than using a dedicated interface→implementation map. This violates the checklist item: "Make Interface Implementations map a class field. You can fill it in using `Map.of()`." Add a class-level Map<Class<?>, Class<?>> (initialized with Map.of(...)) and use it to resolve interface implementations deterministically.

mateacademy-ai-mentor commented on src/main/java/mate.academy/lib/Injector.java:
```diff
@@ -1,13 +1,128 @@
+        if (!CLASS_SET.contains(interfaceClazz)) {
+            throw new UnsupportedOperationException();
+        }
+        Class<?> implementation = findImplementation(interfaceClazz);
```
When a concrete implementation is chosen you do not verify that it is annotated with @Component before instantiation. This violates the checklist instruction: "Don't forget about annotations checks in your Injector class." Per the description, if a class lacks @Component the Injector must refuse to create it and throw an informative exception. Add an explicit check like `if (!implementation.isAnnotationPresent(Component.class)) throw new RuntimeException("Missing @Component on " + implementation)`.

mateacademy-ai-mentor commented on src/main/java/mate.academy/lib/Injector.java:
```diff
@@ -1,13 +1,128 @@
+        Object o;
+        try {
+            o = implementation.getConstructor().newInstance();
+        } catch (InstantiationException | IllegalAccessException
+                 | InvocationTargetException | NoSuchMethodException e) {
```
Reflection-related exceptions are caught and rethrown as `new RuntimeException(e)` without an informative message, and multiple specific reflection exceptions are listed in the catch. This violates: "When throwing an exception add an informative message to it. Also, don't forget to add an exception you're catching in `catch` block to the `RuntimeException` object you throw" and "It is better to replace many exceptions that have a common parent with a general parental exception." Prefer catching `ReflectiveOperationException` and rethrow `new RuntimeException("Injection failed for " + implementation.getName(), e)`.

mateacademy-ai-mentor commented on src/main/java/mate.academy/lib/Injector.java:
```diff
@@ -1,13 +1,128 @@
+        }
+        for (Field field : fields) {
+            if (field.isAnnotationPresent(Inject.class)) {
+                Object instance = getInstance(field.getType());
```
For injected fields you call `getInstance(field.getType())` without first resolving the concrete implementation or verifying the dependency class has @Component. This again violates: "Don't forget about annotations checks in your Injector class." Resolve the dependency class via the interface→implementation map (or check the concrete type) and ensure the resolved class is annotated with @Component before attempting to instantiate or return it.

mateacademy-ai-mentor commented on src/main/java/mate.academy/lib/Injector.java:
```diff
@@ -1,13 +1,128 @@
+                    try {
+                        return url.toURI();
+                    } catch (URISyntaxException e) {
+                        throw new RuntimeException(e);
```
Several utility methods catch checked exceptions (e.g., URISyntaxException) and rethrow `new RuntimeException(e)` without context. The checklist asks: "When throwing an exception add an informative message to it. Also, don't forget to add an exception you're catching in `catch` block to the `RuntimeException` object you throw" — improve these by adding contextual messages (e.g., "Failed to convert URL to URI for resource '...': ") and include the caught exception as the cause.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

I am requesting changes because the current `Injector.java` misses several core requirements that block correct functionality. Specifically:

- No instances cache: you always create new objects without checking a `Map<Class<?>, Object>` first (src/main/java/mate.academy/lib/Injector.java, around instance creation). This violates the single-instance requirement and can worsen circular dependencies. Add a class-level cache, check it before construction, and store newly created instances. [CHECKLIST ITEM #5]
- No interface→implementation map field: implementations are resolved by scanning `COMPONENTS` rather than using a class-level `Map<Class<?>, Class<?>>` initialized via `Map.of(...)`. Add that map and use it deterministically to resolve interfaces. [CHECKLIST ITEM #6]
- Missing @Component checks: before instantiating any chosen class (both the requested type and injected dependencies), verify `implementation.isAnnotationPresent(Component.class)` and throw an informative `RuntimeException` if missing. [CHECKLIST ITEM #4]
- Uninformative exception handling: multiple specific reflection exceptions are caught and rethrown as `new RuntimeException(e)` without context. Prefer catching `ReflectiveOperationException` and rethrow `new RuntimeException("Injection failed for " + implementation.getName(), e)`. Also add context to other catches (e.g., `URISyntaxException`). [CHECKLIST ITEM #7]

Minor: consider basic circular dependency detection and remove debug prints. Your annotations and services look good—once you fix the injector per above, the solution should meet the requirements.

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

koder95 committed:
Update Injector.java

koder95 committed:
Cache interface implementations in Injector

Introduced a map to cache resolved implementations for interface classes in the Injector. This avoids repeated lookups and improves performance when creating instances.

koder95 committed:
Refactor import statements in Injector.java

Replaces wildcard import of java.util.* with explicit imports for Arrays, HashMap, Map, Objects, and Set to improve code clarity and maintainability.

mateacademy-ai-mentor commented on src/main/java/mate.academy/lib/Injector.java:
```diff
@@ -1,13 +1,144 @@
+            .filter(c -> c.isAnnotationPresent(Component.class))
+            .collect(Collectors.toSet());
+
+    private final Map<Class<?>, Class<?>> implementations = new HashMap<>();
```
This field is intended to be the interface→implementation bindings, but the checklist requires a class-level map initialized (e.g., via Map.of(...)) for deterministic resolution. Quote: "Provide a class-level map of interface-to-implementation bindings (e.g., Map<Class<?>, Class<?>>) as a field of the Injector, and initialize it (e.g., using Map.of(...)) [CHECKLIST ITEM #6]." Initialize this map with known bindings instead of populating ad-hoc at runtime.

mateacademy-ai-mentor commented on src/main/java/mate.academy/lib/Injector.java:
```diff
@@ -1,13 +1,144 @@
+            implementation = implementations.get(interfaceClazz);
+        }
+        Field[] fields = implementation.getDeclaredFields();
+        Object o;
+        try {
+            o = implementation.getConstructor().newInstance();
```
An instances cache is missing. The checklist requires maintaining and consulting a cache (instances map) before creating a new instance to avoid redundant creations and to help with dependency graphs. Quote: "The Injector must maintain and consult a cache (instances map) before creating a new instance to avoid redundant creations and help with dependency graphs [CHECKLIST ITEM #5]." Add a Map<Class<?>, Object> instances cache, check it at the start of getInstance, and store newly created objects there.

mateacademy-ai-mentor commented on src/main/java/mate.academy/lib/Injector.java:
```diff
@@ -1,13 +1,144 @@
+            o = implementation.getConstructor().newInstance();
+        } catch (InstantiationException | IllegalAccessException
+                 | InvocationTargetException | NoSuchMethodException e) {
+            throw new RuntimeException(e);
```
Reflection-based instantiation errors are caught and rethrown as `new RuntimeException(e)` without context. The checklist requires informative messages and preferring catching the reflective parent exception. Quote: "Error handling: when rethrowing exceptions, include an informative message and attach the caught exception as the cause; prefer catching the common parent ReflectiveOperationException for reflection-related failures [CHECKLIST ITEM #7 and #8]." Catch ReflectiveOperationException (or at least wrap with `new RuntimeException("Injection failed for " + implementation.getName(), e)`) and include the original exception as the cause.

mateacademy-ai-mentor commented on src/main/java/mate.academy/lib/Injector.java:
```diff
@@ -1,13 +1,144 @@
+    }
+
     public Object getInstance(Class<?> interfaceClazz) {
-        return null;
+        if (!CLASS_SET.contains(interfaceClazz)) {
+            throw new UnsupportedOperationException();
+        }
+        Class<?> implementation;
+        if (!implementations.containsKey(interfaceClazz)) {
+            implementation = findImplementation(interfaceClazz);
+            implementations.put(interfaceClazz, implementation);
+        } else {
+            implementation = implementations.get(interfaceClazz);
+        }
```
Before instantiating or returning an implementation you must verify that the implementation class is annotated with @Component and, if not, throw an informative RuntimeException. Quote: "The Injector must verify annotations before instantiation: check presence of @Component on the implementation class, otherwise throw an informative RuntimeException [CHECKLIST ITEM #4]." Currently there is no such check (and `UnsupportedOperationException` is thrown earlier without a helpful message). Add explicit checks and descriptive exception messages for both requested types and their dependencies.

mateacademy-ai-mentor commented on src/main/java/mate.academy/lib/Injector.java:
```diff
@@ -1,13 +1,144 @@
+    }
+
+    private static Class<?> findImplementation(Class<?> interfaceClazz) {
+        System.out.println("Finding implementation for: " + interfaceClazz);
+        if (interfaceClazz.isInterface()) {
+            return COMPONENTS.stream()
+                    .filter(c -> Arrays.asList(c.getInterfaces()).contains(interfaceClazz))
+                    .findFirst().orElseThrow();
```
`findImplementation` resolves implementations by scanning COMPONENTS at runtime. The checklist requires deterministic resolution via a predefined map instead of ad-hoc scanning. Quote: "The Injector should deterministically resolve interfaces to implementations using a predefined map, not by ad-hoc scanning [CHECKLIST ITEM #6]." Replace this ad-hoc search with lookups into the predefined interface→implementation map.

mateacademy-ai-mentor commented on src/main/java/mate.academy/lib/Injector.java:
```diff
@@ -1,13 +1,144 @@
+                .map(Path::toString)
+                .map(path -> path.substring(path.indexOf(parentPath.split("/")[0])))
+                .map(path -> path.replaceAll("[\\\\/]", "."))
+                .peek(System.out::println)
```
This `peek(System.out::println)` call prints debugging output during class discovery. Please remove debug prints (they clutter output and tests). This was also requested in previous corrections (medium priority).

mateacademy-ai-mentor commented on src/main/java/mate.academy/lib/Injector.java:
```diff
@@ -1,13 +1,144 @@
+    }
+
+    private static Class<?> findImplementation(Class<?> interfaceClazz) {
+        System.out.println("Finding implementation for: " + interfaceClazz);
```
This `System.out.println` is a debug print in `findImplementation`. Remove debug output and rely on proper logging or exceptions instead (debug prints interfere with deterministic test outputs).

mateacademy-ai-mentor commented on src/main/java/mate.academy/lib/Injector.java:
```diff
@@ -1,13 +1,144 @@
+                    try {
+                        return Path.of(url.toURI());
+                    } catch (URISyntaxException e) {
+                        throw new RuntimeException(e);
```
Several places (e.g., Path.of(url.toURI()) catch blocks) rethrow `new RuntimeException(e)` without context. The checklist requires informative messages and including original exception as the cause. Quote: "Error handling: when rethrowing exceptions, include an informative message and attach the caught exception as the cause; prefer catching the common parent ReflectiveOperationException for reflection-related failures [CHECKLIST ITEM #7 and #8]." Wrap these exceptions with meaningful messages (e.g., which resource or classpath failed) and pass the original exception as the cause.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

I’m requesting changes because the Injector violates several mandatory checklist items that affect correctness and determinism.

Critical issues to fix:
- Add an instances cache: `private static final Map<Class<?>, Object> instances = new ConcurrentHashMap<>();` and check it at the start of `getInstance` before creating new objects; store newly created instances there [CHECKLIST ITEM #5].
- Provide a predefined, class-level interface→implementation binding map (e.g., `Map.of(...)`) and resolve implementations using this map rather than scanning at runtime; remove the ad-hoc COMPONENTS scanning in `findImplementation` [CHECKLIST ITEM #6].
- Before instantiation, verify the implementation class is annotated with `@Component`; if not, throw an informative `RuntimeException` indicating the requested type and missing annotation [CHECKLIST ITEM #4].
- Improve exception handling: catch `ReflectiveOperationException` where applicable and rethrow with a clear message like `"Injection failed for " + implementation.getName()` while attaching the original cause; also replace generic `new RuntimeException(e)` elsewhere with meaningful messages (e.g., which resource/classpath failed) [CHECKLIST ITEM #7, #8].
- Remove debug prints (`peek(System.out::println)`, `System.out.println(...)`) to avoid non-deterministic output during tests.

Minor/nice-to-have:
- Consider basic circular dependency detection using a construction stack to produce clearer errors.

Your annotations and service implementations look good—great job there. Please address the Injector fixes above to meet the core requirements and ensure deterministic, safe injection.

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

koder95 committed:
Improve error messages and remove debug prints in Injector

Enhanced RuntimeException messages for better debugging and removed unnecessary System.out.println debug statements from the Injector class.

koder95 committed:
Update Injector.java

slade13 commented on src/main/java/mate.academy/lib/Injector.java:
```diff
@@ -1,13 +1,142 @@
+        Object o;
+        try {
+            o = implementation.getConstructor().newInstance();
+        } catch (InstantiationException | IllegalAccessException
```
You catch `InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException`. Prefer a single catch (`ReflectiveOperationException e`) per the checklist.

slade13 reviewed (approved):
Don't use `main` branch for your PR's. Always create a new feature branch. Next time I will reject such PR.

koder95 commented:
Ok. I didn't notice this rule while doing any of the tasks. Thank you for give me the notice and accept the task. I am grateful and I obligate send solutions in my own branch.
